File: src/confirm.ts
export type ConfirmDecision = "ok" | "ask" | "require-yes";

export function confirmDecision(options: {
  clone: boolean;
  dryRun: boolean;
  yes: boolean;
  tty: boolean;
}): ConfirmDecision {
  if (options.clone || options.dryRun) return "ok";
  if (options.yes) return "ok";
  if (options.tty) return "ask";
  return "require-yes";
}

export function isTty(
  stdin: { isTTY?: boolean } = process.stdin,
  stdout: { isTTY?: boolean } = process.stdout,
): boolean {
  return Boolean(stdin.isTTY && stdout.isTTY);
}

export function requireYesMessage(command: "install" | "uninstall"): string {
  return `non-interactive ${command} requires --yes
  incodex ${command} --yes`;
}

File: crates/incodex-cli/src/confirm.rs
use std::io::{self, Write};

pub const CONFIRM_PROMPT: &str = "Press Enter to confirm, ESC to cancel: ";

pub fn ask_to_continue() -> Result<bool, String> {
    print!("{CONFIRM_PROMPT}");
    io::stdout().flush().map_err(|err| err.to_string())?;
    #[cfg(not(target_os = "windows"))]
    {
        let key = crate::terminal::read_key()?;
        println!();
        return match key.as_slice() {
            [3] => Err("interrupted".into()),
            [b'\r'] | [b'\n'] => Ok(true),
            _ => Ok(false),
        };
    }
    #[cfg(target_os = "windows")]
    {
        use crate::menu_controller::MenuKey;

        let key = crate::windows_console::read_menu_key()?;
        println!();
        match key {
            MenuKey::Activate => Ok(true),
            MenuKey::Interrupt => Err("interrupted".into()),
            _ => Ok(false),
        }
    }
}

pub(crate) fn require(command: &str, yes: bool) -> Result<(), String> {
    if yes {
        return Ok(());
    }
    #[cfg(not(target_os = "windows"))]
    let interactive = crate::terminal::is_tty();
    #[cfg(target_os = "windows")]
    let interactive = crate::windows_console::is_tty();
    if interactive {
        return if ask_to_continue()? {
            Ok(())
        } else {
            Err("aborted".into())
        };
    }
    Err(format!(
        "non-interactive {command} requires --yes\n  incodex {command} --yes"
    ))
}
